_posts/2020/2020-08-10-spring-boot-upgrade-not-allowed.md

```diff
@@ -0,0 +1,124 @@
+Our proxy is built on top of [Zuul](https://github.com/Netflix/zuul){:target="_blank"}, which was sending the needed headers correctly.
+Emphasis on *was*.
+
+## Why does it have to be beans?
+The problem is simple.
+We need forwarded headers.
+We also need Spring Boot version upgrade.
+What do we do?
+
+It's the Bean time!
+```java
+@Bean
+public Filter forwardedHeaderFilter() {
+    return new ForwardedHeaderFilter();
+}
+```
+
+With this bean in configuration class I run the tests again, and they all successfully pass.
+The proxy plays well with the server behind the proxy, but how do other microservices play with the proxy itself?
+
+We have debug environments for such cases, so that you can quickly check if the changes are working as expected, and also run functional API tests while you are at it.
+I go straight into tests and run all of them, fully expecting them to be green after they finish.
+The tests start one by one and each successfully completes.
+Business, as usual.
+This goes on for some time until one test suddenly fails.
+Then another.
+And another.
+And a dozen more.
+All of them are related to uploading and managing the images on the storage.
+
+And now the vicious loop begins.
+Searching for similar issues, trying out the solutions from all the different posts on Github and Stackoverflow, failing to run them successfully, searching for new solutions, etc.
+Nothing helped.
+My team tried to lend me a hand and after every discussion I would go and make a change in the microservice, only to find out that the changes did not solve the problem.
+
+After a certain point I did not believe that we can upgrade our proxy microservice in the first place and felt like giving up on the task altogether (which means putting it back to the backlog).
+This was when Donald told me he can take this problem over.
+
+---
+
+Integration tests, Distributed Tracing and Logging are the terms that we developers tend to ignore or not worry that much since they don't appear in the scene on sunny days.
+But working on a microservice architecture where quite often you have to debug requests passing through several services, the lack of traceability or a centralized logging system can be quite a hassle.
+Luckily our logs are analyzed and stored to `Google Stackdriver` which offers you plenty of functionalities to search through the logs.
+
+Let's go back to the problem and to what I did to find it and come up with the solution in the end.
+
+After discussing with Andrey what had he tried and I decide to go Sherlock and follow the failing requests around in order to see what was changed on its way.
+
+{% image_custom image="/assets/img/pages/blog/images/sherlock.png" width="80" lightbox %}
+
+As it's know every request in a Spring based environment goes through a chain of filters which can validate, enhance or even block the requests.
```

Comment:
As it is known, every request in a Spring based environment goes through a chain of filters which can validate, enhance, or even block the requests.